File: backend/app/models/app_settings.py
# Copyright (c) 2025 Laurent Barbe
# Licensed under the Apache License, Version 2.0
import re
from typing import Optional
from urllib.parse import urlparse

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.core.config import get_settings
from app.models.storage_endpoint import StorageEndpointPublic


def _default_portal_cors_origins() -> list[str]:
    return list(get_settings().cors_origins or [])


def _default_bucket_migration_parallelism_max() -> int:
    return get_settings().bucket_migration_parallelism_max


def _default_bucket_migration_max_active_per_endpoint() -> int:
    return get_settings().bucket_migration_max_active_per_endpoint


class PortalBucketDefaultsOverride(BaseModel):
    model_config = ConfigDict(extra="forbid")

    versioning: Optional[bool] = None
    enable_cors: Optional[bool] = None
    enable_lifecycle: Optional[bool] = None
    noncurrent_version_expiration_days: Optional[int] = Field(default=None, ge=1)
    cors_allowed_origins: Optional[list[str]] = None


class PortalSettingsOverride(BaseModel):
    model_config = ConfigDict(extra="forbid")

    browser_access_enabled: Optional[bool] = None
    allow_private_storage_space_create: Optional[bool] = None
    allow_portal_named_bucket_create: Optional[bool] = None
    allow_portal_user_access_key_create: Optional[bool] = None
    server_access_logging_enabled: Optional[bool] = None
    storage_space_version_cleanup_enabled: Optional[bool] = None
    bucket_defaults: Optional[PortalBucketDefaultsOverride] = None


class PortalSettingsAdminUpdate(PortalSettingsOverride):
    delegated_to_portal_managers: Optional[bool] = None


class PortalBucketDefaults(BaseModel):
    versioning: bool = True
    enable_cors: bool = True
    enable_lifecycle: bool = True
    noncurrent_version_expiration_days: int = Field(default=90, ge=1)
    cors_allowed_origins: list[str] = Field(default_factory=_default_portal_cors_origins)


class GeneralSettings(BaseModel):
    manager_enabled: bool = True
    ceph_admin_enabled: bool = False
    storage_ops_enabled: bool = False
    browser_enabled: bool = True
    browser_root_enabled: bool = True
    browser_manager_enabled: bool = False
    browser_portal_enabled: bool = True
    browser_ceph_admin_enabled: bool = False
    portal_enabled: bool = False
    billing_enabled: bool = False
    endpoint_status_enabled: bool = True
    quota_alerts_enabled: bool = False
    usage_history_enabled: bool = True
    bucket_migration_enabled: bool = False
    bucket_compare_enabled: bool = True
    bucket_integrity_check_enabled: bool = True
    bucket_usage_stats_enabled: bool = True
    bucket_purge_enabled: bool = False
    bucket_quota_management_enabled: bool = True
    manager_ceph_s3_user_keys_enabled: bool = True
    allow_login_access_keys: bool = False
    allow_login_endpoint_list: bool = False
    allow_login_custom_endpoint: bool = False


class GeneralFeatureLock(BaseModel):
    forced: bool = False
    value: Optional[bool] = None
    source: Optional[str] = None


class GeneralFeatureLocks(BaseModel):
    manager_enabled: GeneralFeatureLock = Field(default_factory=GeneralFeatureLock)
    ceph_admin_enabled: GeneralFeatureLock = Field(default_factory=GeneralFeatureLock)
    storage_ops_enabled: GeneralFeatureLock = Field(default_factory=GeneralFeatureLock)
    browser_enabled: GeneralFeatureLock = Field(default_factory=GeneralFeatureLock)
    portal_enabled: GeneralFeatureLock = Field(default_factory=GeneralFeatureLock)
    billing_enabled: GeneralFeatureLock = Field(default_factory=GeneralFeatureLock)
    endpoint_status_enabled: GeneralFeatureLock = Field(default_factory=GeneralFeatureLock)


class BrandingSettings(BaseModel):
    primary_color: str = "#0ea5e9"
    login_logo_url: Optional[str] = None

    @field_validator("primary_color", mode="before")
    @classmethod
    def normalize_primary_color(cls, value: Optional[str]) -> str:
        if value is None:
            return "#0ea5e9"
        if not isinstance(value, str):
            raise ValueError("primary_color must be a string")
        normalized = value.strip().lower()
        if not normalized:
            return "#0ea5e9"
        if not re.fullmatch(r"#[0-9a-f]{6}", normalized):
            raise ValueError("primary_color must be a hex color in #rrggbb format")
        return normalized

    @field_validator("login_logo_url", mode="before")
    @classmethod
    def normalize_login_logo_url(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        if not isinstance(value, str):
            raise ValueError("login_logo_url must be a string")
        normalized = value.strip()
        if not normalized:
            return None
        if normalized.startswith("/"):
            return normalized
        if normalized.startswith("data:image/"):
            return normalized
        parsed = urlparse(normalized)
        if parsed.scheme in {"http", "https"} and bool(parsed.netloc):
            return normalized
        raise ValueError("login_logo_url must be http(s), root-relative (/...), or data:image/... URL")


class LoginSettings(BaseModel):
    allow_login_access_keys: bool = False
    allow_login_endpoint_list: bool = False
    allow_login_custom_endpoint: bool = False
    default_endpoint_url: Optional[str] = None
    endpoints: list[StorageEndpointPublic] = Field(default_factory=list)
    login_logo_url: Optional[str] = None
    seed_login_prefill: bool = False
    seed_login_email: Optional[str] = None
    seed_login_password: Optional[str] = None


class PortalSettings(BaseModel):
    model_config = ConfigDict(extra="forbid")

    browser_access_enabled: bool = False
    allow_private_storage_space_create: bool = True
    allow_portal_named_bucket_create: bool = False
    allow_portal_user_access_key_create: bool = True
    server_access_logging_enabled: bool = True
    server_access_log_retention_days: int = Field(default=30, ge=1)
    storage_space_version_cleanup_enabled: bool = True
    max_portal_user_access_keys: int = Field(default=2, ge=1)
    bucket_defaults: PortalBucketDefaults = Field(default_factory=PortalBucketDefaults)


class ManagerSettings(BaseModel):
    manager_rgw_usage_metrics_enabled: bool = True
    bucket_migration_parallelism_default: int = Field(
        default_factory=_default_bucket_migration_parallelism_max,
        ge=1,
        le=128,
    )
    bucket_migration_parallelism_max: int = Field(
        default_factory=_default_bucket_migration_parallelism_max,
        ge=1,
        le=128,
    )
    bucket_migration_max_active_per_endpoint: int = Field(
        default_factory=_default_bucket_migration_max_active_per_endpoint,
        ge=1,
        le=64,
    )

    @model_validator(mode="after")
    def validate_bucket_migration_limits(self):
        if self.bucket_migration_parallelism_default > self.bucket_migration_parallelism_max:
            raise ValueError("bucket_migration_parallelism_default must be <= bucket_migration_parallelism_max")
        return self


class QuotaNotificationSettings(BaseModel):
    threshold_percent: int = Field(default=85, ge=1, le=100)
    include_subject_contact_email: bool = False
    smtp_host: Optional[str] = None
    smtp_port: int = Field(default=587, ge=1, le=65535)
    smtp_username: Optional[str] = None
    smtp_from_email: Optional[str] = None
    smtp_from_name: Optional[str] = None
    smtp_starttls: bool = True
    smtp_timeout_seconds: int = Field(default=15, ge=1, le=300)

    @field_validator("smtp_host", "smtp_username", "smtp_from_email", "smtp_from_name", mode="before")
    @classmethod
    def normalize_optional_strings(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        if not isinstance(value, str):
            raise ValueError("SMTP settings must be strings")
        normalized = value.strip()
        return normalized or None


class BrowserSettings(BaseModel):
    allow_proxy_transfers: bool = True
    direct_upload_parallelism: int = Field(default=5, ge=1, le=20)
    proxy_upload_parallelism: int = Field(default=2, ge=1, le=20)
    direct_download_parallelism: int = Field(default=5, ge=1, le=20)
    proxy_download_parallelism: int = Field(default=2, ge=1, le=20)
    other_operations_parallelism: int = Field(default=3, ge=1, le=20)
    streaming_zip_threshold_mb: int = Field(default=200, ge=0, le=10240)


class OnboardingSettings(BaseModel):
    dismissed: bool = False


class AppSettings(BaseModel):
    general: GeneralSettings = Field(default_factory=GeneralSettings)
    portal: PortalSettings = Field(default_factory=PortalSettings)
    manager: ManagerSettings = Field(default_factory=ManagerSettings)
    quota_notifications: QuotaNotificationSettings = Field(default_factory=QuotaNotificationSettings)
    browser: BrowserSettings = Field(default_factory=BrowserSettings)
    onboarding: OnboardingSettings = Field(default_factory=OnboardingSettings)
    branding: BrandingSettings = Field(default_factory=BrandingSettings)

File: frontend/src/api/appSettings.ts
/*
 * Copyright (c) 2025 Laurent Barbe
 * Licensed under the Apache License, Version 2.0
 */
import client from "./client";

export type PortalSettings = {
  allow_portal_key: boolean;
  allow_portal_user_bucket_create: boolean;
  allow_portal_user_access_key_create: boolean;
  iam_group_manager_policy: PortalIAMPolicySettings;
  iam_group_user_policy: PortalIAMPolicySettings;
  bucket_access_policy: PortalIAMPolicySettings;
  bucket_defaults: PortalBucketDefaults;
  override_policy: PortalSettingsOverridePolicy;
};

export type GeneralSettings = {
  manager_enabled: boolean;
  ceph_admin_enabled: boolean;
  storage_ops_enabled: boolean;
  browser_enabled: boolean;
  browser_root_enabled: boolean;
  browser_manager_enabled: boolean;
  browser_portal_enabled: boolean;
  browser_ceph_admin_enabled: boolean;
  allow_portal_manager_workspace: boolean;
  portal_enabled: boolean;
  billing_enabled: boolean;
  endpoint_status_enabled: boolean;
  quota_alerts_enabled: boolean;
  usage_history_enabled: boolean;
  bucket_migration_enabled: boolean;
  bucket_compare_enabled: boolean;
  allow_ui_user_bucket_migration: boolean;
  allow_login_access_keys: boolean;
  allow_login_endpoint_list: boolean;
  allow_login_custom_endpoint: boolean;
  allow_user_private_connections: boolean;
};

export type GeneralFeatureLock = {
  forced: boolean;
  value?: boolean | null;
  source?: string | null;
};

export type GeneralFeatureLocks = {
  manager_enabled: GeneralFeatureLock;
  ceph_admin_enabled: GeneralFeatureLock;
  storage_ops_enabled: GeneralFeatureLock;
  browser_enabled: GeneralFeatureLock;
  portal_enabled: GeneralFeatureLock;
  billing_enabled: GeneralFeatureLock;
  endpoint_status_enabled: GeneralFeatureLock;
};

export type BrandingSettings = {
  primary_color: string;
  login_logo_url: string | null;
};

export type PortalIAMPolicySettings = {
  actions: string[];
  advanced_policy?: Record<string, unknown> | null;
};

export type PortalIAMPolicyOverridePolicy = {
  actions: boolean;
  advanced_policy: boolean;
};

export type PortalBucketDefaultsOverridePolicy = {
  versioning: boolean;
  enable_cors: boolean;
  enable_lifecycle: boolean;
  cors_allowed_origins: boolean;
};

export type PortalSettingsOverridePolicy = {
  allow_portal_key: boolean;
  allow_portal_user_bucket_create: boolean;
  allow_portal_user_access_key_create: boolean;
  iam_group_manager_policy: PortalIAMPolicyOverridePolicy;
  iam_group_user_policy: PortalIAMPolicyOverridePolicy;
  bucket_access_policy: PortalIAMPolicyOverridePolicy;
  bucket_defaults: PortalBucketDefaultsOverridePolicy;
};

export type PortalBucketDefaults = {
  versioning: boolean;
  enable_cors: boolean;
  enable_lifecycle: boolean;
  cors_allowed_origins: string[];
};

export type PortalIAMPolicyOverride = {
  actions?: string[] | null;
  advanced_policy?: Record<string, unknown> | null;
};

export type PortalBucketDefaultsOverride = {
  versioning?: boolean | null;
  enable_cors?: boolean | null;
  enable_lifecycle?: boolean | null;
  cors_allowed_origins?: string[] | null;
};

export type PortalSettingsOverride = {
  allow_portal_key?: boolean | null;
  allow_portal_user_bucket_create?: boolean | null;
  allow_portal_user_access_key_create?: boolean | null;
  iam_group_manager_policy?: PortalIAMPolicyOverride | null;
  iam_group_user_policy?: PortalIAMPolicyOverride | null;
  bucket_access_policy?: PortalIAMPolicyOverride | null;
  bucket_defaults?: PortalBucketDefaultsOverride | null;
};

export type ManagerSettings = {
  allow_manager_user_usage_stats: boolean;
  bucket_migration_parallelism_default: number;
  bucket_migration_parallelism_max: number;
  bucket_migration_max_active_per_endpoint: number;
};

export type QuotaNotificationSettings = {
  threshold_percent: number;
  include_subject_contact_email: boolean;
  smtp_host?: string | null;
  smtp_port: number;
  smtp_username?: string | null;
  smtp_from_email?: string | null;
  smtp_from_name?: string | null;
  smtp_starttls: boolean;
  smtp_timeout_seconds: number;
};

export type BrowserSettings = {
  allow_proxy_transfers: boolean;
  direct_upload_parallelism: number;
  proxy_upload_parallelism: number;
  direct_download_parallelism: number;
  proxy_download_parallelism: number;
  other_operations_parallelism: number;
  streaming_zip_threshold_mb: number;
};

export type OnboardingSettings = {
  dismissed: boolean;
};

export type AppSettings = {
  general: GeneralSettings;
  portal: PortalSettings;
  manager: ManagerSettings;
  quota_notifications: QuotaNotificationSettings;
  browser: BrowserSettings;
  onboarding: OnboardingSettings;
  branding: BrandingSettings;
};

export type QuotaNotificationTestEmailResult = {
  status: string;
  recipient: string;
  sent_at: string;
};

export type PublicStorageEndpoint = {
  id: number;
  name: string;
  endpoint_url: string;
  is_default: boolean;
};

export type LoginSettings = {
  allow_login_access_keys: boolean;
  allow_login_endpoint_list: boolean;
  allow_login_custom_endpoint: boolean;
  default_endpoint_url?: string | null;
  endpoints: PublicStorageEndpoint[];
  login_logo_url?: string | null;
  seed_login_prefill?: boolean;
  seed_login_email?: string | null;
  seed_login_password?: string | null;
};

export async function fetchAppSettings(): Promise<AppSettings> {
  const { data } = await client.get<AppSettings>("/admin/settings");
  return data;
}

export async function fetchDefaultAppSettings(): Promise<AppSettings> {
  const { data } = await client.get<AppSettings>("/admin/settings/defaults");
  return data;
}

export async function fetchGeneralFeatureLocks(): Promise<GeneralFeatureLocks> {
  const { data } = await client.get<GeneralFeatureLocks>("/admin/settings/general-feature-locks");
  return data;
}

export async function fetchGeneralSettings(): Promise<GeneralSettings> {
  const { data } = await client.get<GeneralSettings>("/settings/general");
  return data;
}

export async function fetchLoginSettings(): Promise<LoginSettings> {
  const { data } = await client.get("/settings/login");
  const normalized = (data && typeof data === "object" ? data : {}) as Partial<LoginSettings>;
  return {
    allow_login_access_keys: Boolean(normalized.allow_login_access_keys ?? false),
    allow_login_endpoint_list: Boolean(normalized.allow_login_endpoint_list ?? false),
    allow_login_custom_endpoint: Boolean(normalized.allow_login_custom_endpoint ?? false),
    default_endpoint_url: normalized.default_endpoint_url ?? null,
    endpoints: Array.isArray(normalized.endpoints) ? normalized.endpoints : [],
    login_logo_url: typeof normalized.login_logo_url === "string" ? normalized.login_logo_url : null,
    seed_login_prefill: Boolean(normalized.seed_login_prefill ?? false),
    seed_login_email: normalized.seed_login_email ?? null,
    seed_login_password: normalized.seed_login_password ?? null,
  };
}

export async function fetchBrandingSettings(): Promise<BrandingSettings> {
  const { data } = await client.get<BrandingSettings>("/settings/branding");
  const normalized = (data && typeof data === "object" ? data : {}) as Partial<BrandingSettings>;
  return {
    primary_color: typeof normalized.primary_color === "string" ? normalized.primary_color : "#0ea5e9",
    login_logo_url: typeof normalized.login_logo_url === "string" ? normalized.login_logo_url : null,
  };
}

export async function updateAppSettings(payload: AppSettings): Promise<AppSettings> {
  const { data } = await client.put<AppSettings>("/admin/settings", payload);
  return data;
}

export async function sendQuotaNotificationTestEmail(
  payload: QuotaNotificationSettings
): Promise<QuotaNotificationTestEmailResult> {
  const { data } = await client.post<QuotaNotificationTestEmailResult>("/admin/settings/quota-notifications/test-email", payload);
  return data;
}
